### DISCUSSION TITLE
Volume and thickness constraints for airfoil case

### DISCUSSION THREAD

**Discussion by FaiqShahbaz** [mdolab/dafoam#708] [General]
In the tutorial the volume and thickness constraints were applied using telist and lelist, from which a straight line is drawn and divided in 10 segments. But what if the airfoil is not symmetric and has a very less camber, the same approach will result in error since that line will be outside the actual airfoil.
How to deal with this kind of situation? 
![image_2024-11-20_214740195](https://github.com/user-attachments/assets/4be4fa3b-1c8a-40b3-9a74-d8346b9278c9)

i have set up constraints as below,
        # setup the volume and thickness constraints
        leList = [[1e-4, 0.0, 1e-4], [1e-4, 0.0, 0.1 - 1e-4]]
        teList = [[ 0.992735 - 1e-4, 0.0017 - 1e-4, 1e-4], [0.998 - 1e-4, 0.002712 - 1e-4, 0.1 - 1e-4]]
where the last points of airfoil are as,
PS
0.968141  0.005908
0.982146  0.003613
0.992735  0.001700
SS
0.973434  0.010465
0.984387  0.006072
0.992735  0.002712

but since the the airfoil profile is unsymmetric as you can see the straight line wont work and there will be error projecting node.

![airfoil_profile](https://github.com/user-attachments/assets/dd93ff29-608e-4730-b5f6-419f5630c85d)

**Comment by friedenhe** [comment]
In this case, you can put your leList and teList completely outside of the airfoil, for example, at y=-0.1. When pyGeo projects in the y+ direction, it will find two intersection points on the airfoil surface, it will use these two points for computing the thickness volume.

**Reply by FaiqShahbaz** [reply]
i tried putting le and telist outside of the airfoil at y=-0.1 but its giving error.
# volume and thickness constraints
        leList = [[1e-4, -0.1, 1e-4], [1e-4, -0.1, 0.1 - 1e-4]]
        teList = [[ 0.992735 - 1e-4, -0.1, 1e-4], [0.998 - 1e-4, -0.1, 0.1 - 1e-4]]
        
# Error
 +------------------------------------------------------------------------------+
| Error: There was an error projecting a node at (0.992635, -0.100000,         |
| 0.000100) with normal (0.000000, 0.011006, 0.000000).                        |
+------------------------------------------------------------------------------+

**Comment by friedenhe** [comment]
1e-4 and 0.9927 are too close to the LE and TE and may mess up the projection, try to move them away from TE and LE.

**Reply by FaiqShahbaz** [reply]
i am experiencing the same error, i have tried multiple point till -0.1 (for le) and 1.1( for te) away from le and te and yet in the end experiencing projection error, is it possible that the error is entirely something else.

**Reply by friedenhe** [reply]
You should use x from ~0.1 to ~0.9
